app/data_sources/mattermost.py

```diff
@@ -0,0 +1,163 @@
+    @staticmethod
+    def get_config_fields() -> List[ConfigField]:
+        return [
+            ConfigField(label="Mattermost Server", name="url", placeholder="https://mattermost.server.com", input_type=HTMLInputType.TEXT),
+            ConfigField(label="Token", name="token", placeholder="paste-your-token-here", input_type=HTMLInputType.PASSWORD),
+        ]
+
+
+    @staticmethod
+    def validate_config(config: Dict) -> None:
+        try:
+            parsed_config = MattermostConfig(**config)
+            maattermost = Driver(options = asdict(parsed_config))
+            maattermost.login()
+        except Exception as e:
+            raise InvalidDataSourceConfig from e
+
+    
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        mattermost_config = MattermostConfig(**self._config)
+        self._mattermost = Driver(options = asdict(mattermost_config))
+    
+    
+    def _list_channels(self) -> List[MattermostChannel]:
+        channels = self._mattermost.channels.client.get("/channels")
+        return [MattermostChannel(id=channel["id"], name=channel["name"], team_id=channel["team_id"])
+                for channel in channels]
+
+
+    def _is_valid_message(self, message: Dict) -> bool:
+        return message["type"] == ""
+    
+    
+    def _list_posts_in_channel(self, channel_id: str, page: int) -> Dict:
+        endpoint = f"/channels/{channel_id}/posts"
+        params = {
+            "since": int(self._last_index_time.timestamp()),
+            "page": page
+        }
+        
+        posts = self._mattermost.channels.client.get(endpoint, params=params)
+        return posts
+
+
+    def _feed_new_documents(self) -> None:
+        self._mattermost.login()
+        channels = self._list_channels()
+    
+        logger.info(f'Found {len(channels)} channels')
+        parse_with_workers(self._parse_channel_worker, channels)
+
+    
+    def _parse_channel_worker(self, channels: List[MattermostChannel]):        
+        for channel in channels:
+            self._feed_channel(channel)
+
+    
+    def _get_mattermost_url(self):
+        options = self._mattermost.options
+        return f"{options['scheme']}://{options['url']}:{options['port']}"
+
+
+    def _get_team_url(self, channel: MattermostChannel):
+        url = self._get_mattermost_url()
+        return f"{url}/{channel.team_id}"
+        
+
+    def _feed_channel(self, channel: MattermostChannel):
+        logger.info(f'Feeding channel {channel.name}')
+        
+        page = 0
+        total_fed = 0
+        
+        parsed_posts = []
+        
+        team_url = self._get_team_url(channel)
+        while True:
+            posts = self._list_posts_in_channel(channel.id, page)
+
+            for id, post in posts["posts"].items():
+                if not self._is_valid_message(post):
+                    continue
+                
+                author = self._mattermost.users.get_user(post["user_id"])["username"]
```

Comment:
Cache mechanism is required here 
look at google drive, Bary used a nice builtin decorator for that

Reply:
Done